File: go/mechanisms/evm/exact/client/erc20_approval.go
package client

import (
	"context"
	"encoding/hex"
	"fmt"
	"math/big"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/core/types"

	"github.com/coinbase/x402/go/extensions/erc20approvalgassponsor"
	"github.com/coinbase/x402/go/mechanisms/evm"
)

// SignErc20ApprovalTransaction creates a signed (but unbroadcast) EIP-1559 transaction
// that calls approve(Permit2, MaxUint256) on the given ERC-20 token.
//
// The returned Info contains the RLP-encoded signed transaction which the facilitator
// will broadcast before calling settle().
func SignErc20ApprovalTransaction(
	ctx context.Context,
	signer evm.ClientEvmSignerWithTxSigning,
	tokenAddress string,
	chainID *big.Int,
) (*erc20approvalgassponsor.Info, error) {
	owner := signer.Address()
	normalizedToken := evm.NormalizeAddress(tokenAddress)
	spender := evm.PERMIT2Address

	// Encode calldata: approve(spender, MaxUint256)
	contractABI, err := abi.JSON(strings.NewReader(string(evm.ERC20ApproveABI)))
	if err != nil {
		return nil, fmt.Errorf("failed to parse approve ABI: %w", err)
	}
	maxUint256 := evm.MaxUint256()
	calldata, err := contractABI.Pack("approve", common.HexToAddress(spender), maxUint256)
	if err != nil {
		return nil, fmt.Errorf("failed to encode approve calldata: %w", err)
	}

	// Get nonce
	nonce, err := signer.GetTransactionCount(ctx, owner)
	if err != nil {
		return nil, fmt.Errorf("failed to get transaction count: %w", err)
	}

	// EstimateFeesPerGas returns usable fallback values even on RPC error,
	// but guard against nil to avoid panic in DynamicFeeTx construction.
	maxFeePerGas, maxPriorityFeePerGas, feeErr := signer.EstimateFeesPerGas(ctx)
	if feeErr != nil && (maxFeePerGas == nil || maxPriorityFeePerGas == nil) {
		return nil, fmt.Errorf("failed to estimate fees and no fallback available: %w", feeErr)
	}

	// Build EIP-1559 transaction
	toAddr := common.HexToAddress(normalizedToken)
	tx := types.NewTx(&types.DynamicFeeTx{
		ChainID:   chainID,
		Nonce:     nonce,
		GasTipCap: maxPriorityFeePerGas,
		GasFeeCap: maxFeePerGas,
		Gas:       evm.ERC20ApproveGasLimit,
		To:        &toAddr,
		Value:     big.NewInt(0),
		Data:      calldata,
	})

	// Sign the transaction
	rlpBytes, err := signer.SignTransaction(ctx, tx)
	if err != nil {
		return nil, fmt.Errorf("failed to sign approve transaction: %w", err)
	}

	signedTxHex := "0x" + hex.EncodeToString(rlpBytes)

	return &erc20approvalgassponsor.Info{
		From:              owner,
		Asset:             normalizedToken,
		Spender:           spender,
		Amount:            maxUint256.String(),
		SignedTransaction: signedTxHex,
		Version:           erc20approvalgassponsor.ERC20ApprovalGasSponsoringVersion,
	}, nil
}

File: typescript/packages/mechanisms/evm/src/exact/client/erc20approval.ts
import { encodeFunctionData, getAddress, maxUint256 } from "viem";
import {
  PERMIT2_ADDRESS,
  erc20ApproveAbi,
  ERC20_APPROVE_GAS_LIMIT,
  DEFAULT_MAX_FEE_PER_GAS,
  DEFAULT_MAX_PRIORITY_FEE_PER_GAS,
} from "../../constants";
import { ClientEvmSigner } from "../../signer";
import {
  ERC20_APPROVAL_GAS_SPONSORING_VERSION,
  type Erc20ApprovalGasSponsoringInfo,
} from "../extensions";

export type Erc20ApprovalTxSigner = Pick<ClientEvmSigner, "address"> & {
  signTransaction: NonNullable<ClientEvmSigner["signTransaction"]>;
  getTransactionCount: NonNullable<ClientEvmSigner["getTransactionCount"]>;
  estimateFeesPerGas?: NonNullable<ClientEvmSigner["estimateFeesPerGas"]>;
};

/**
 * Signs an EIP-1559 `approve(Permit2, MaxUint256)` transaction for the given token.
 *
 * The signed transaction is NOT broadcast here — the facilitator broadcasts it
 * atomically before settling the Permit2 payment. This enables Permit2 payments
 * for generic ERC-20 tokens that do NOT implement EIP-2612.
 *
 * Always approves MaxUint256 regardless of the payment amount.
 *
 * @param signer - The client EVM signer (must support signTransaction, getTransactionCount)
 * @param tokenAddress - The ERC-20 token contract address
 * @param chainId - The chain ID
 * @returns The ERC-20 approval gas sponsoring info object
 */
export async function signErc20ApprovalTransaction(
  signer: Erc20ApprovalTxSigner,
  tokenAddress: `0x${string}`,
  chainId: number,
): Promise<Erc20ApprovalGasSponsoringInfo> {
  const from = signer.address;
  const spender = getAddress(PERMIT2_ADDRESS);

  // Encode approve(PERMIT2_ADDRESS, MaxUint256) calldata
  const data = encodeFunctionData({
    abi: erc20ApproveAbi,
    functionName: "approve",
    args: [spender, maxUint256],
  });

  // Get current nonce for the sender
  const nonce = await signer.getTransactionCount({ address: from });

  // Get current fee estimates, with fallback values
  let maxFeePerGas: bigint;
  let maxPriorityFeePerGas: bigint;
  try {
    const fees = await signer.estimateFeesPerGas?.();
    if (!fees) {
      throw new Error("no fee estimates available");
    }
    maxFeePerGas = fees.maxFeePerGas;
    maxPriorityFeePerGas = fees.maxPriorityFeePerGas;
  } catch {
    maxFeePerGas = DEFAULT_MAX_FEE_PER_GAS;
    maxPriorityFeePerGas = DEFAULT_MAX_PRIORITY_FEE_PER_GAS;
  }

  // Sign the EIP-1559 transaction (not broadcast)
  const signedTransaction = await signer.signTransaction({
    to: tokenAddress,
    data,
    nonce,
    gas: ERC20_APPROVE_GAS_LIMIT,
    maxFeePerGas,
    maxPriorityFeePerGas,
    chainId,
  });

  return {
    from,
    asset: tokenAddress,
    spender,
    amount: maxUint256.toString(),
    signedTransaction,
    version: ERC20_APPROVAL_GAS_SPONSORING_VERSION,
  };
}
